### PR TITLE
Correct function type for vtable shim (fix some boxed closures)

### LINKED ISSUE #367: Correct function type for vtable shim
Breaking off from https://github.com/model-checking/rmc/issues/240 to have distinct issues.

The problem is that we fail verification for calls through `FnOnce` vtable closures:
```
let f: Box<dyn FnOnce()> = Box::new(|x| {
    assert(x == 2); // should fail
});
f(1);
```

### PR DESCRIPTION
**Pull request by avanhatt**
### Description of changes: 

RMC currently fails to verify boxed dynamic `FnOnce` closures, like this:

```rust
let f: Box<dyn FnOnce()> = Box::new(|x| {
    assert(x == 2); // should fail
});
f(1);
```

The issue is somewhat specific to `FnOnce::call_once`, which takes a `self` pointer instead of a `self *` pointer at the source language level. Vtable calls expect a pointer as the first argument, so Rust itself repairs vtable shims of this type to take an argument:

```rust
    /// `<T as Trait>::method` where `method` receives unsizeable `self: Self` (part of the
    /// `unsized_locals` feature).
    ///
    /// The generated shim will take `Self` via `*mut Self` - conceptually this is `&owned Self` -
    /// and dereference the argument to call the original function.
    VtableShim(DefId),

...
if let ty::InstanceDef::VtableShim(..) = self.def {
    // Modify `fn(self, ...)` to `fn(self: *mut Self, ...)`.
    sig = sig.map_bound(|mut sig| {
        let mut inputs_and_output = sig.inputs_and_output.to_vec();
        inputs_and_output[0] = tcx.mk_mut_ptr(inputs_and_output[0]);
        sig.inputs_and_output = tcx.intern_type_list(&inputs_and_output);
        sig
    });
}
```

We need to do the same fix.

### Resolved issues: 

Resolves #367

### Call-outs:

Known remaining issues: https://github.com/model-checking/rmc/issues/240

Broader concerns are tracked in: https://github.com/model-checking/rmc/issues/299.

### Testing:

* How is this change tested?

- Existing unit test (fixes 2 tests that previously required disabling `--pointer-check`).
- New `_fixme` test to be fixed in #240 

* Is this a refactor change?

No.

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: compiler/rustc_codegen_llvm/src/gotoc/place.rs [hunk 1]
```diff
@@ -208,26 +208,26 @@ impl<'tcx> GotocCtx<'tcx> {
...
-                _ => None,
-            },
+            // A local can be pointer to a FnDef, like Fn::call and Fn::call_mut
+            ty::RawPtr(inner) => self.codegen_local_fndef(inner.ty).map(|f| f.address_of()),
+            // A local can be a boxed function pointer
+            ty::Adt(def, _) if def.is_box() => {
+                let boxed_ty = self.codegen_ty(ty);
+                self.codegen_local_fndef(ty.boxed_ty())
+                    .map(|f| self.box_value(f.address_of(), boxed_ty))
+            }
```
Comment (adpaco-aws) [inline compiler/rustc_codegen_llvm/src/gotoc/place.rs:222]: Can you update the function-level comment regarding this case?

File: src/test/cbmc/DynTrait/boxed_closure.rs [hunk 2]
```diff
@@ -1,13 +1,6 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0 OR MIT
 
-//rmc-flags: --no-memory-safety-checks
-
-// We use `--no-memory-safety-checks` in this test to avoid getting
-// a verification failure:
-// [pointer_dereference.7] invalid function pointer: FAILURE
-// Tracking issue: https://github.com/model-checking/rmc/issues/307
```
Comment (adpaco-aws) [inline src/test/cbmc/DynTrait/boxed_closure.rs:9]: Let's not forget to mark this one as "done" in #307 once this PR is merged.

**Comment by avanhatt** [comment]
Found a new test that still fails, converting to draft while I work on the fix.

File: compiler/rustc_codegen_llvm/src/gotoc/place.rs [hunk 3]
```diff
@@ -208,26 +208,26 @@ impl<'tcx> GotocCtx<'tcx> {
     /// code for Fn::call and Fn::call_mut both use pointers to a FnDef.
     /// In these cases, we need to generate an expression that references the
     /// existing fndef rather than a named variable.
-    pub fn codegen_local_fndef(&mut self, l: Local) -> Option<Expr> {
-        let t = self.local_ty(l);
-        match t.kind() {
+    pub fn codegen_local_fndef(&mut self, ty: &'tcx ty::TyS<'tcx>) -> Option<Expr> {
```
[the remark this answers is not in the document]
Comment (avanhatt) [reply compiler/rustc_codegen_llvm/src/gotoc/place.rs:211]: Added `_fixme` test; but actual fix to support it is involved and will be split off into a different PR.

File: compiler/rustc_codegen_llvm/src/gotoc/place.rs (hunk 1 above)
Comment (avanhatt) [reply compiler/rustc_codegen_llvm/src/gotoc/place.rs:225]: Done!

**Review by danielsn — APPROVED** [review]

**Pull request merged**